Create packages lock file for packages.config projects on install and uninstall

zivkan opened the pull request:
NuGet/Home#7951
spec: https://github.com/NuGet/Home/wiki/Packages.config-SHA-Validation

Here's a sample lock file:
```json
{
  "version": 1,
  "dependencies": {
    ".NETFramework,Version=v4.7.2": {
      "Newtonsoft.Json": {
        "type": "Direct",
        "requested": "[12.0.1, 12.0.1]",
        "resolved": "12.0.1",
        "contentHash": "pBR3wCgYWZGiaZDYP+HHYnalVnPJlpP1q55qvVb+adrDHmFMDc1NAKio61xTwftK3Pw5h7TZJPJEEVMd6ty8rg=="
      }
    }
  }
}
```

zivkan committed:
Create packages lock file for packages.config projects on install and uninstall

jainaashish commented on src/NuGet.Core/NuGet.Commands/PackagesLockFileBuilder.cs:
```diff
@@ -89,5 +94,56 @@ public PackagesLockFile CreateNuGetLockFile(LockFile assetsFile)
+                var target = new PackagesLockFileTarget
+                {
+                    TargetFramework = targetFramework.Key,
+                    Dependencies = new List<LockFileDependency>()
```
Dependencies is always initialized to empty list so you don't need to do this.

jainaashish commented on src/NuGet.Core/NuGet.Commands/PackagesLockFileBuilder.cs:
```diff
@@ -89,5 +94,56 @@ public PackagesLockFile CreateNuGetLockFile(LockFile assetsFile)
+                    var file = getNupkgFilePath(package.PackageIdentity);
+                    if (file == null)
+                    {
+                        throw new Exception("couldn't find package");
```
Multiple concerns here:
Should it really throw at this point?
Should it really throw a generic Exception? without any relevant details? Resource string?

zivkan replied:
oops, I intended this to be temporary and to "do it properly" before commiting, but I forgot :(  As part of fixing other issues, my code no longer throws any exceptions.

jainaashish commented on src/NuGet.Core/NuGet.Commands/PackagesLockFileBuilder.cs:
```diff
@@ -89,5 +94,56 @@ public PackagesLockFile CreateNuGetLockFile(LockFile assetsFile)
+
+                    using (var reader = new PackageArchiveReader(file))
+                    {
+                        hash = reader.GetContentHashForSignedPackage(token);
```
I'd imagine by the time you call this API to generate lock file, packages have been installed so there'd be `.nupkg.metadata` file already present for each package. Please use that to get the content hash instead of regenerating that again.

jainaashish commented on src/NuGet.Core/NuGet.Commands/PackagesLockFileBuilder.cs:
```diff
@@ -89,5 +94,56 @@ public PackagesLockFile CreateNuGetLockFile(LockFile assetsFile)
+                    {
+                        Id = package.PackageIdentity.Id,
+                        Type = PackageDependencyType.Direct,
+                        RequestedVersion = package.AllowedVersions,
```
In packages.config world, `RequestedVersion` and `ResolvedVersion` would point to same metadata `packageIdentity.Version`

zivkan replied:
I decided to remove the line, since RequestedVersion won't be read during packages.config restore.

jainaashish replied:
For the sake of consistency, I'd say you keep both requestedVersion as well as resolvedVersion in the lock file so that tomorrow anyone crawling this lock file gets consistent behavior for PC or PR.

jainaashish commented on src/NuGet.Core/NuGet.Commands/PackagesLockFileBuilder.cs:
```diff
@@ -89,5 +94,56 @@ public PackagesLockFile CreateNuGetLockFile(LockFile assetsFile)
+
+            var lockFile = new PackagesLockFile();
+
+            foreach (var targetFramework in installedPackages.GroupBy(p => p.TargetFramework).OrderBy(g => g.Key))
```
what happens when there is no `targetFramework` on installed packages in `Packages.config`? How are you categorizing those?

zivkan replied:
The class that reads packages.config sets the TargetFramework to `Unsupported,V0.0`. I can map this to `Any,V0.0` in the builder if Unsupported looks too ugly in the lock file.

jainaashish replied:
`Any` makes more sense to me.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2358,69 @@ var projectUniqueNamesForBuildIntToUpdate
                         }
                     }
 
+#if PACKAGE_CONFIG_LOCK
```
Why do you need this if/def? where is this defined?

zivkan replied:
I originally thought to merge into dev, and I didn't want dev builds to use these changes until the feature is ready. I was setting it on my machine only. But since I created a feature branch, it's no longer necessary.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2358,69 @@ var projectUniqueNamesForBuildIntToUpdate
+                        if (msbuildProject != null)
+                        {
+                            //start batch processing for msbuild
+                            await msbuildProject.ProjectSystem.BeginProcessingAsync();
```
you don't need this batch processing here.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2358,69 @@ var projectUniqueNamesForBuildIntToUpdate
+                            //todo: get lock file name same way PackageReference projects do
+                            var directory = Path.GetDirectoryName(msbuildProject.MSBuildProjectPath);
+                            var lockFileFullPath = Path.Combine(directory, PackagesLockFileFormat.LockFileName);
+                            var lockFileName = Path.GetFileName(lockFileFullPath);
```
you can use `PackagesLockFileUtilities.GetNuGetLockFilePath` or add an override api there for your purpose. But this will completely ignore custom lock file names.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2358,69 @@ var projectUniqueNamesForBuildIntToUpdate
+                            var lockFileName = Path.GetFileName(lockFileFullPath);
+
+                            var installedPackagesEnumeration = await nuGetProject.GetInstalledPackagesAsync(token);
+                            var installedPackages = installedPackagesEnumeration as ICollection<PackageReference> ?? installedPackagesEnumeration.ToList();
```
why did you need to type cast it to `ICollection`?

zivkan replied:
I was doing it because Linq's `.ToList()` is expensive, particularly for large lists. If object is an ICollection we can cast much more cheaply than creating a new list and possibly resizing multiple times.

jainaashish replied:
why do you even need to convert it to `List`? just pass the `IEnumerable` down the line...

zivkan replied:
The original code needed to know if the result was empty or not. Multiple enumerations of IEnumerable throws exceptions (for example calling .Any() followed by a foreach), depending on data source, so I avoid doing anywhere as a rule.

However, I changed the code to keep the lock file when there are no packages installed. It creates a lock file without any targets or packages. I did this to ensure if that if someone opts-in to lock files, then uninstalls all packages and then installs a package, they're still opted-in to lock files. So, it no longer needs different code paths for zero packages, so this cast/check is no longer done.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2358,69 @@ var projectUniqueNamesForBuildIntToUpdate
+                            {
+                                if (File.Exists(lockFileFullPath))
+                                {
+                                    File.Delete(lockFileFullPath);
```
instead use `FileSystemUtility.DeleteFile`

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2358,69 @@ var projectUniqueNamesForBuildIntToUpdate
+                        {
+                            // end batch for msbuild and let it save everything.
+                            // always calls it before PostProcessAsync or binding redirects
+                            await msbuildProject.ProjectSystem.EndProcessingAsync();
```
don't need this

jainaashish commented on test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests/PackagesConfigProjectTests/PackagesLockTests.cs:
```diff
@@ -0,0 +1,158 @@
+        public async Task DoNotCreateLockFileWhenFeatureDisabled()
+        {
+            // Arrange
+            var sourceRepositoryProvider = TestSourceRepositoryUtility.CreateV3OnlySourceRepositoryProvider();
```
don't use http source instead create local source to run your tests. 

I know some of the existing tests are still pointing to V3 sources but all new tests are being added with their own local source and clean afterwards.

jainaashish reviewed (changes requested):
It's going in the right direction.. some suggestions for improvements.

zivkan committed:
Improvements for PR comments

- Prefer .nupkg.metadata to get content hash. Fallback to other methods if .nupkg.metadata file doesn't exist.
- Remove #if PACKAGE_CONFIG_LOCK since feature branch is being used instead of merging to dev
- get lock file name from project properties, if set

zivkan committed:
More PR feedback changes

- packages without a targetFramework should read "Any", not "Unsupported", in the lock file
- Specify RequestedVersion as the specific version installed

jainaashish commented on src/NuGet.Core/NuGet.Commands/Utility/ContentHashUtility.cs:
```diff
@@ -0,0 +1,154 @@
+
+namespace NuGet.Commands.Utility
+{
+    public class ContentHashUtility : IContentHashUtility
```
I' not sure about the purpose of this whole file... As I mentioned in my earlier review, you should just be able to use `.nupkg.metadata` file to get the content hash for each package from global packages folder. Only exception will be when someone used `DirectDownload` to install the package but that also isn't applicable for your current scenario (within VS). 

So it should be pretty straight forward, am I missing anything here?

zivkan replied:
If I clear my global packages cache after installing a package, when I install a second package, the first package will not be re-extracted into the global packages folder, so the .nupkg.metadata file will not exist.

Let's talk offline.

jainaashish replied:
Still not sure why do you need this? didn't we discuss that you only update lock file incrementally instead of regenerating on every install?

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2359,55 @@ var projectUniqueNamesForBuildIntToUpdate
+                    var directory = Path.GetDirectoryName(msbuildProject.MSBuildProjectPath);
+                    var msbuildProperty = msbuildProject?.ProjectSystem?.GetPropertyValue("NuGetLockFilePath");
+                    string lockFileName = null;
+                    if (msbuildProperty is string nuGetLockFilePathValue)
```
nit: space before if condition

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2359,55 @@ var projectUniqueNamesForBuildIntToUpdate
+                            lockFileName = Path.Combine(directory, nuGetLockFilePathValue);
+                        }
+                    }
+                    if (lockFileName == null)
```
nit: space

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2358,6 +2359,55 @@ var projectUniqueNamesForBuildIntToUpdate
+                    if (useLockFile || File.Exists(lockFileName))
+                    {
+                        var installedPackagesEnumeration = await nuGetProject.GetInstalledPackagesAsync(token);
+                        var installedPackages = installedPackagesEnumeration as ICollection<PackageReference> ?? installedPackagesEnumeration.ToList();
```
forgot to remove this line? and passing the enumeration to generate lock file

jainaashish reviewed (changes requested):
My main concern is with getting content hash for each package. We can discuss it offline and decides there.

zivkan committed:
Modify PC lock file instead of regenerating it

jainaashish commented:
@zivkan ping team whenever its ready to review again.

zivkan commented:
@jainaashish @nkolev92 @PatoBeltran  This is ready for another round of reviews. It runs without any test failures on my local machine, but the CI build is broken :(

jainaashish commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -205,12 +206,16 @@ private Task AddFileCoreAsync(string path, Action addFile)
+                && !fileName.Equals(EnvDTEProjectInfoUtility.AppConfig)
+                && !fileName.Equals(PackagesLockFileFormat.LockFileName)
+                && !fileName.Equals("packages." + ProjectName + ".lock.json")
+                && !(customLockFilePath != null && fileName.Equals(customLockFilePath)))
```
there would always be only one file name so you don't need to check for every combination. There is already an utility function to get the packages lock file path.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,158 @@
+            var lockFileName = GetPackagesLockFilePath(msbuildProject);
+            var lockFileExists = File.Exists(lockFileName);
+            var enableLockFile = IsPackagesLockFileExcplicitlyEnabled(msbuildProject);
+            if (enableLockFile || lockFileExists)
```
if lock file exists but EnableLockFile is false then it's an error case for us. So you'd need to handle this

zivkan replied:
I pushed a commit that adds this error condition, but I haven't tested it yet. Still, if you have a better suggestion than the exception that I'm currently throwing, I'll certainly take it on board.

zivkan replied:
The exception appears in the error list, but without an error code. I haven't yet investigated how to have the error code appear  correctly.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,158 @@
+            {
+                Debug.Assert(toUninstall.NuGetProjectActionType == NuGetProjectActionType.Uninstall);
+
+                foreach (var installed in lockFile.Targets[0].Dependencies)
```
why it's always being removed from first target? what happens when there are multi-targets?

zivkan replied:
PC lock files will always use the "Any,V0.0" TFM. Partly because it's easier (the actions list doesn't specify which TFM the package should use, so we would have to get additional information from elsewhere to use the 'actual' TFM). Partly because PC projects can't multi-target and doesn't have the capability to have conditional references depending on target framework, like PR can. The PC file is a flat a flat list, so the lock file can be too.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,158 @@
+
+                foreach (var installed in lockFile.Targets[0].Dependencies)
+                {
+                    if (installed.Id == toUninstall.PackageIdentity.Id)
```
since package is is ignore case, so use string comparison with ignore case option

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,158 @@
+            {
+                Debug.Assert(toUninstall.NuGetProjectActionType == NuGetProjectActionType.Uninstall);
+
+                foreach (var installed in lockFile.Targets[0].Dependencies)
```
you can even simplified this foreach with a linq query

zivkan replied:
is
```csharp
var toRemove = lockFile.Targets[0].Dependencies.FirstOrDefault(d => string.Equals(d.Id, toUninstall.PackageIdentity.Id, StringComparison.InvariantCultureIgnoreCase));
if (toRemove != null)
{
    lockFile.Targets[0].Dependencies.Remove(toRemove);
}
```
simpler than
```csharp
foreach (var installed in lockFile.Targets[0].Dependencies)
{
    if (string.Equals(installed.Id, toUninstall.PackageIdentity.Id, StringComparison.InvariantCultureIgnoreCase))
    {
        lockFile.Targets[0].Dependencies.Remove(installed);
        return;
    }
}
```
? It's less lines, but it's a much longer line

There would be a minor performance gain if we avoid using IEnumerable and LINQ (and the delegate, anonymous function and closure that comes with LINQ)

```csharp
var toUninstallId = toUninstall.PackageIdentity.Id;
var dependencies = lockFile.Targets[0].Dependencies;
for (var i = 0; i < dependencies.Count; i++)
{
    if (string.Equals(toUninstallId,  lockFile.Targets[0].Dependencies[i].Id, StringComparison.InvariantCultureIgnoreCase))
    {
        dependencies.RemoveAt(i);
        break;
    }
}
```

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,158 @@
+                    Type = PackageDependencyType.Direct
+                };
+
+                // should keep sorted, but lockFile[Targets[0].Dependencies is an IList<T>, but only List<T> has the .Sort() method, so we have to insert in sorted order.
```
why do you need to keep them in sorted order? current impl will really make it hard to see the file diff since it will add new packages at random places which might mess with uninstalling packages.

So I'd suggest to keep it simple to keep appending at the end unless there is a valid reason to do otherwise

zivkan replied:
The PC file is sorted, so it keeps the lock file consistent. And the reason I think it's a good idea is exactly to simplify file diffs. Since the file is JSON, adding a new item at the end will cause 2 lines to diff, the new line and adding a comma to the end of the last item. Also, if a user removes a package, but then installs it again, I don't want their lock file to be different. Keeping it sorted makes sure that the lock file only has a source control diff when the contents are the same, no matter which order they were installed.

jainaashish reviewed (changes requested):
It looks good, just few concerns and suggestions.

zivkan committed:
Get lock file from utility, rather than whitelist all possible values

Also compare package ids with case-insensitive comparer.

zivkan commented:
I'm working on this PR again, so please hold off final reviews for now. thanks.

zivkan committed:
Changes due to design change

- PC lock file target framework should be project's target framework
- Read content hash from solution packages folder .nupkg.metadata file
- Write content hash to solution packages folder when it doesn't exist, only when content hash needed (so packages that don't need sha validation don't get metadata files)

various bug fixes while testing:

- when lock file is empty, after reading the Version was Int.MinValue. Make sure we always write the current file format version
- handle case when lock file is empty

zivkan commented:
This PR is ready for review again.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -205,12 +207,14 @@ private Task AddFileCoreAsync(string path, Action addFile)
             // it into the project.
             // Other exceptions are 'web.config' and 'app.config'
             var fileName = Path.GetFileName(path);
+            var lockFileFullPath = PackagesConfigLockFileUtility.GetPackagesLockFilePath(ProjectFullPath, GetPropertyValue("NuGetLockFilePath")?.ToString(), ProjectName);
```
Do we have any other places where we read settings from the csproj in PackageReference?
Other than "where" the packages.config file is that is. 
 Just trying to understand if we are adding a new gesture.

zivkan replied:
`MSBuildProjectSystem.TargetFramework` reads various target framework properties, but that's the only other case which I can find. Although that surprises me, because packing a `csproj` should read some properties, right? Or a `nuspec` that has `$variables$`.

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/VsMSBuildProjectSystem.cs:
```diff
@@ -205,12 +207,14 @@ private Task AddFileCoreAsync(string path, Action addFile)
                 && !fileName.Equals(EnvDTEProjectInfoUtility.WebConfig)
-                && !fileName.Equals(EnvDTEProjectInfoUtility.AppConfig))
+                && !fileName.Equals(EnvDTEProjectInfoUtility.AppConfig)
+                && !fileName.Equals(Path.GetFileName(lockFileFullPath)))
```
crazy nit: to avoid editing unintentional lines every time a new option is added, I prefer having the closing element on a new line.

nkolev92 commented on src/NuGet.Core/NuGet.Commands/Utility/SolutionPackagesContentHashUtility.cs:
```diff
@@ -0,0 +1,159 @@
+        {
+            // try to read the .nupkg.metadata file from the solution packages folder
+            var nupkgPath = GetNupkgPath(packageIdentity, _packagesFolderSourceRepository, token);
+            var result = TryNupkgMetadata(nupkgPath);
```
nit: naming - TryGetNupkgMetadata maybe?

nkolev92 commented on src/NuGet.Core/NuGet.Commands/Utility/SolutionPackagesContentHashUtility.cs:
```diff
@@ -0,0 +1,159 @@
+                var globalNupkgPath = GetNupkgPathInGlobalPackagesFolder(packageIdentity, token);
+                if (globalNupkgPath != null)
+                {
+                    result = TryNupkgMetadata(globalNupkgPath);
```
Analogous to our discussion on Monday, i have concern whether this is a good approach. 
You cannot guarantee that the nupkg in the solution packages folder is the same as the nupkg in the global packages folder. 

Furthermore what happens if the nupkg is in the solution packages folder but not in the global packages folder?

nkolev92 commented on src/NuGet.Core/NuGet.Commands/Utility/SolutionPackagesContentHashUtility.cs:
```diff
@@ -0,0 +1,159 @@
+
+        private string GetNupkgPath(PackageIdentity packageIdentity, SourceRepository source, CancellationToken token)
+        {
+            var findPackages = source.GetResource<FindLocalPackagesResource>();
```
Is the GetContentHashAsync method invoked N-times where N is the number of packages?
Since there are only 2 possible sources, I'd cache the resource to avoid creating too much garbage.

nkolev92 commented on src/NuGet.Core/NuGet.Commands/Utility/SolutionPackagesContentHashUtility.cs:
```diff
@@ -0,0 +1,159 @@
+            }
+
+            _packagesFolderSourceRepository = packagesFolderSourceRepository;
+            _globalPackageFolderRepositories = globalPackageFolderRepositories;
```
global package folder repositorie**s**.
Multiple? 
How?

zivkan replied:
fallback folders were in the list when I was debugging. I'm not sure if it adds all local folders, or if the fallback folder was somehow special, but it was there.

nkolev92 commented on src/NuGet.Core/NuGet.PackageManagement/Strings.resx:
```diff
@@ -491,4 +491,8 @@
     <comment>0 - new line separated list of errors</comment>
   </data>
+  <data name="Error_InvalidLockFileInput" xml:space="preserve">
+    <value>Invalid restore input where RestorePackagesWithLockFile property is set to false and also exist a packages lock file at {0}.</value>
```
the RestorePackage... and exist**s** ?

zivkan replied:
I copied this error message from another assembly. Should I fix it there too?

`also exist a packages lock file` sounds really weird to me. I'd like to change it to `also a packages locak file exists` in both assemblies, but not sure if we want to cause the translators extra work.

nkolev92 reviewed (commented):
Are you uising the same PackagesLockFile model as PR?

Can you please post a sample of the how the PC lock file looks?

nkolev92 commented on src/NuGet.Core/NuGet.Commands/Utility/SolutionPackagesContentHashUtility.cs:
```diff
@@ -0,0 +1,159 @@
+            _logger = logger;
+        }
+
+        public async Task<string> GetContentHashAsync(PackageIdentity packageIdentity, CancellationToken token)
```
discussed offline, these should be internal.

zivkan committed:
More code review changes.

- Move the new content hash utility to NuGet.PackageManagement so that it can be made internal
- Stop falling back to the global packages folder to look for a `.nupkg.metadata` file, since the solution package might have been restored from a different source and therefore have a different content hash.
- Improve error message string

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/ISolutionPackagesContentHashUtility.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace NuGet.PackageManagement
+{
+    internal interface ISolutionPackagesContentHashUtility
```
why there is a interface for utility class? are there or going to be multiple implementation for this?

zivkan replied:
so it can be mocked and tests that use the utility don't need nupkgs on disk, assuming the method is otherwise testable.

jainaashish replied:
Then why even include `Utility` in the class name... felt lil odd to me. 

Just a suggestion to rename it as `IPackagesContentHashProvider`… but you're free to ignore.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,205 @@
+        internal static async Task UpdateLockFileAsync(
+            MSBuildNuGetProject msbuildProject,
+            List<NuGetProjectAction> actionsList,
+            SourceRepository packagesFolderSourceRepository,
```
you don't need this. Just take `MSBuildNuGetProject` instance that will give you instance of `FolderNuGetProject`. Use that to extract content hash of the packages

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,205 @@
+            else if (enableLockFile == true || lockFileExists)
+            {
+                var lockFile = GetLockFile(lockFileExists, lockFileName);
+                lockFile.Targets[0].TargetFramework = msbuildProject.ProjectSystem.TargetFramework;
```
`packages.config` can have multiple packages targeting different set of framework which are compatible with each other, like look at nuget repo file `.nuget\packages.config` file. So it's quite possible to have multiple target frameworks in lock file even for `packages.config` based projects.

Now, I'm not sure if ignoring these frameworks is the right thing? This seems reasonable, but be careful and check some of these multi framework scenarios manually to make sure it works fine in all cases.

zivkan replied:
This repo's `packages.config` says `Newtonsoft.Json` should use `targetFramework="net45"`. However, when I look at my `packages\Newtonsoft.Json.9.0.1\lib` folder, I see all 7 TFMs from the nupkg (and yes, they contain dlls), not just net45.

Additionally when I looked for usage of `targetFramework`, I could only find writes, not reads. Therefore I don't believe that NuGet actually reads/uses the `targetFramework` in the `packages.config` file. For this reason, I don't believe it should be considered valid to list the same package two or more times in a single `packages.config` file, even if they have different `targetFramework`s.

jainaashish replied:
I'm just pointing this out to keep in mind. If you've already verified then great.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,205 @@
+                else if (lockFile.Targets.Count > 1)
+                {
+                    // merge lists
+                    while (lockFile.Targets.Count > 1)
```
Aren't you always creating this lock file with single framework? when will it have multi frameworks?

zivkan replied:
it's just for safety in case someone hand-edits the file or uses some unofficial tool to generate it.

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackagesConfigLockFileUtility.cs:
```diff
@@ -0,0 +1,205 @@
+                    var projectUri = new Uri(msbuildProject.MSBuildProjectPath);
+                    var lockFileUri = new Uri(lockFileName);
+                    var lockFileRelativePath = projectUri.MakeRelativeUri(lockFileUri).OriginalString;
+                    if (Path.DirectorySeparatorChar != '/')
```
why do you need to calculate lockFileRelativePath? You can just pass the `lockFileName` or `Path.GetFileName(lockFileName)` to `AddExistingFile` and it should work.

zivkan replied:
I tried passing the absolute path (`lockFileName`), but there's some kind of bug that makes VS try to copy my entire c: drive into the project and add every file as an item to the project. `Path.GetFileName(lockFileName)` will fail if the customer uses `LockFileNamePath` to put the lock file in a directory other than where the csproj is (either a sub-directory or a parent directory).

jainaashish commented on src/NuGet.Core/NuGet.PackageManagement/Utility/SolutionPackagesContentHashUtility.cs:
```diff
@@ -0,0 +1,121 @@
+{
+    internal class SolutionPackagesContentHashUtility : ISolutionPackagesContentHashUtility
+    {
+        private readonly FindLocalPackagesResource _packagesFolderFindPackagesResource;
```
Instead use `FolderNuGetProject`

jainaashish reviewed (commented):
This looks very good. Few suggestions regarding utility interface and using `FolderNuGetProject` instead of `SourceRepository` to get content hash of the packages.

jainaashish reviewed (approved):
There are few open suggestions to further improve the code readability and efficiency, otherwise this looks good.

zivkan committed:
code review improvements

- rename content hash class and interface
- Use FolderNuGetProject instead of SourceRepository/FindLocalPackagesResource.

zivkan committed:
Skip PC lock file checking when there is no MSBuildNuGetProject

zivkan committed:
fix compile errors after rebase

nkolev92 commented:
@zivkan 
Can we get an example of wha the lock file looks like? 
And maybe a link to the issue and spec in the PR?

zivkan committed:
fix tests

zivkan committed:
Fix bug for website projects (no msbuild project file)
